system/services/admin/DataManagerBatchOperationService.cfc

```diff
@@ -0,0 +1,429 @@
+				}
+				if ( canWarn ) {
+					arguments.logger.warn( $translateResource( uri="cms:datamanager.batchdelete.task.no.records.deleted.message" ) );
+				}
+				return true;
+			}
+			if ( canReportProgress && !arguments.batchAll ) {
+				arguments.progress.setProgress( arguments.audit ? 50 : 90 );
+			}
+			if ( canInfo ) {
+				arguments.logger.info( $translateResource( uri="cms:datamanager.batchdelete.task.deleted.records.message", data=[ objectTitle, NumberFormat( deletedCount ) ] ) );
+			}
+
+			// audit
+			if ( arguments.audit ) {
+				for( var record in args.records ) {
+					$audit(
+						  action   = arguments.auditAction
+						, type     = arguments.auditType
+						, userId   = arguments.auditUserId
+						, recordId = record.id
+						, detail   = { id=record.id, label=record.label, objectName=arguments.objectName }
+					);
+
+					if ( canReportProgress && !arguments.batchAll ) {
+						arguments.progress.setProgress( 50 + Int( ( 50/totalRecords ) * ++audited ) );
+					}
+				}
+			}
+
+			// post delete hooks
+			if ( customizationService.objectHasCustomization( arguments.objectName, "postBatchDeleteRecordsAction" ) ) {
+				customizationService.runCustomization(
+					  objectName = arguments.objectName
+					, action     = "postBatchDeleteRecordsAction"
+					, args       = args
+				);
+			}
+
+			// finish up
+			if ( !moreToFetch ) {
+				if ( canInfo ) {
+					arguments.logger.info( $translateResource( uri="cms:datamanager.batchdelete.task.finished" ) );
+				}
+				if ( canReportProgress ) {
+					arguments.progress.setProgress( 100 );
+				}
+			}
+
+			if ( canReportProgress && moreToFetch ) {
+				batchProgress += args.records.recordCount;
+				arguments.progress.setProgress( ( 100 / totalRecords ) * batchProgress );
+			}
+		} while( moreToFetch );
+
+		return true;
+	}
+
+	public numeric function getBatchSourceRecordCount( required string objectName, required struct sourceArgs ) {
+		return $getPresideObjectService().selectData(
+			  argumentCollection = arguments.sourceArgs
+			, objectName         = arguments.objectName
+			, recordCountOnly    = true
+		);
+	}
+
+	public string function queueBatchOperation(
+		  required string objectName
+		, required struct batchSrcArgs
+	) {
+		var pobjService   = $getPresideObjectService();
+		var queueDataArgs = StructCopy( arguments.batchSrcArgs );
+		var dbAdapter     = pobjService.getDbAdapterForObject( arguments.objectName );
+		var idField       = dbAdapter.escapeEntity( pobjService.getIdField( arguments.objectName ) );
+		var queueId       = CreateUUId();
+
+		queueDataArgs.selectFields = [
+			  "'#queueId#'"
+			, idField
+			, dbAdapter.getNowFunctionSql()
+		];
+
+		pobjService.insertDataFromSelect(
+			  objectName     = "batch_operation_queue"
+			, fieldList      = [ "queue_id", "record_id", "datecreated" ]
+			, selectDataArgs = queueDataArgs
+		);
+
+		return queueId;
+	}
+
+	public numeric function clearBatchOperationQueue( required string queueId ) {
```

Comment:
I wonder if there should be a message here to say how many records were removed from queue?